File: app/src/main/java/com/example/data/repository/SerialRangeHelper.kt
package com.example.data.repository

data class RangeExpansionResult(
    val isValid: Boolean,
    val quantity: Int,
    val serials: List<String>,
    val errorMessage: String? = null
)

object SerialRangeHelper {
    private const val MAX_RANGE_LIMIT = 2000

    /**
     * Safely parses and expands a sequential range from `fromSerial` to `toSerial`.
     * Supports purely numeric serials (e.g., "100001" to "100050")
     * and structured alphanumeric serials with common prefix and numeric suffix (e.g., "TAG100001" to "TAG100050").
     */
    fun expandRange(fromRaw: String, toRaw: String): RangeExpansionResult {
        val from = fromRaw.trim()
        val to = toRaw.trim()

        if (from.isEmpty() || to.isEmpty()) {
            return RangeExpansionResult(false, 0, emptyList(), "From and To serials cannot be empty")
        }

        if (from == to) {
            return RangeExpansionResult(true, 1, listOf(from))
        }

        // Pure numeric case
        if (from.all { it.isDigit() } && to.all { it.isDigit() }) {
            val fromNum = from.toLongOrNull()
            val toNum = to.toLongOrNull()

            if (fromNum == null || toNum == null) {
                return RangeExpansionResult(false, 0, emptyList(), "Numeric serial numbers exceed valid range limits")
            }

            if (fromNum > toNum) {
                return RangeExpansionResult(false, 0, emptyList(), "From serial ($from) must be less than or equal to To serial ($to)")
            }

            val qty = (toNum - fromNum + 1).toInt()
            if (qty > MAX_RANGE_LIMIT) {
                return RangeExpansionResult(false, 0, emptyList(), "Range exceeds maximum allowable limit of $MAX_RANGE_LIMIT tags (Found: $qty)")
            }

            val padLength = from.length
            val list = ArrayList<String>(qty)
            for (i in fromNum..toNum) {
                list.add(i.toString().padStart(padLength, '0'))
            }
            return RangeExpansionResult(true, qty, list)
        }

        // Alphanumeric structured case with prefix + numeric suffix
        val fromMatch = Regex("^([A-Za-z_-]+)(\\d+)$").find(from)
        val toMatch = Regex("^([A-Za-z_-]+)(\\d+)$").find(to)

        if (fromMatch != null && toMatch != null) {
            val prefixFrom = fromMatch.groupValues[1]
            val suffixFrom = fromMatch.groupValues[2]
            val prefixTo = toMatch.groupValues[1]
            val suffixTo = toMatch.groupValues[2]

            if (prefixFrom.uppercase() != prefixTo.uppercase()) {
                return RangeExpansionResult(false, 0, emptyList(), "Prefix mismatch: '$prefixFrom' vs '$prefixTo'")
            }

            val fromNum = suffixFrom.toLongOrNull()
            val toNum = suffixTo.toLongOrNull()

            if (fromNum == null || toNum == null) {
                return RangeExpansionResult(false, 0, emptyList(), "Invalid numeric suffix in structured serial")
            }

            if (fromNum > toNum) {
                return RangeExpansionResult(false, 0, emptyList(), "From serial number ($suffixFrom) must be <= To serial number ($suffixTo)")
            }

            val qty = (toNum - fromNum + 1).toInt()
            if (qty > MAX_RANGE_LIMIT) {
                return RangeExpansionResult(false, 0, emptyList(), "Range size ($qty) exceeds maximum allowable limit of $MAX_RANGE_LIMIT tags")
            }

            val padLength = suffixFrom.length
            val list = ArrayList<String>(qty)
            for (i in fromNum..toNum) {
                list.add(prefixFrom + i.toString().padStart(padLength, '0'))
            }
            return RangeExpansionResult(true, qty, list)
        }

        return RangeExpansionResult(false, 0, emptyList(), "Serial format cannot be automatically incremented. Please enter individual serials or correct format.")
    }
}

File: src/utils/serialRangeHelper.ts
/**
 * Serial Range Helper
 * Matches Android SerialRangeHelper.kt logic for expanding contiguous FASTag serial ranges.
 */

export interface RangeValidationResult {
  valid: boolean;
  error?: string;
  count: number;
}

export class SerialRangeHelper {
  static validateRange(from: string, to: string, maxAllowed = 1000): RangeValidationResult {
    const cleanFrom = from.trim().toUpperCase();
    const cleanTo = to.trim().toUpperCase();

    if (!cleanFrom || !cleanTo) {
      return { valid: false, error: 'Start and end serials cannot be blank.', count: 0 };
    }

    if (cleanFrom.length !== cleanTo.length) {
      return {
        valid: false,
        error: `Serial length mismatch: '${cleanFrom}' (${cleanFrom.length} chars) vs '${cleanTo}' (${cleanTo.length} chars).`,
        count: 0
      };
    }

    // Pure numeric range
    if (/^\d+$/.test(cleanFrom) && /^\d+$/.test(cleanTo)) {
      try {
        const start = BigInt(cleanFrom);
        const end = BigInt(cleanTo);

        if (start > end) {
          return { valid: false, error: `Start serial (${cleanFrom}) cannot be greater than end serial (${cleanTo}).`, count: 0 };
        }

        const count = Number(end - start + 1n);
        if (count > maxAllowed) {
          return { valid: false, error: `Range size (${count.toLocaleString()}) exceeds maximum allowed batch limit of ${maxAllowed}.`, count };
        }

        return { valid: true, count };
      } catch {
        return { valid: false, error: 'Invalid numeric serial range format.', count: 0 };
      }
    }

    // Alpha-numeric prefix with numeric suffix
    const fromMatch = cleanFrom.match(/^([A-Za-z0-9_-]*?)(\d+)$/);
    const toMatch = cleanTo.match(/^([A-Za-z0-9_-]*?)(\d+)$/);

    if (fromMatch && toMatch) {
      const [, prefixFrom, digitsFrom] = fromMatch;
      const [, prefixTo, digitsTo] = toMatch;

      if (prefixFrom !== prefixTo) {
        return { valid: false, error: `Prefix mismatch: '${prefixFrom}' vs '${prefixTo}'.`, count: 0 };
      }

      if (digitsFrom.length !== digitsTo.length) {
        return { valid: false, error: 'Suffix numeric length mismatch.', count: 0 };
      }

      try {
        const start = BigInt(digitsFrom);
        const end = BigInt(digitsTo);

        if (start > end) {
          return { valid: false, error: `Start number (${digitsFrom}) cannot exceed end number (${digitsTo}).`, count: 0 };
        }

        const count = Number(end - start + 1n);
        if (count > maxAllowed) {
          return { valid: false, error: `Range size (${count.toLocaleString()}) exceeds maximum allowed limit of ${maxAllowed}.`, count };
        }

        return { valid: true, count };
      } catch {
        return { valid: false, error: 'Failed to compute alphanumeric serial range.', count: 0 };
      }
    }

    return { valid: false, error: 'Unrecognized serial format. Must end with numeric sequence.', count: 0 };
  }

  static expandRange(from: string, to: string, maxAllowed = 1000): string[] {
    const validation = this.validateRange(from, to, maxAllowed);
    if (!validation.valid) {
      throw new Error(validation.error || 'Invalid range');
    }

    const cleanFrom = from.trim().toUpperCase();
    const cleanTo = to.trim().toUpperCase();

    if (/^\d+$/.test(cleanFrom) && /^\d+$/.test(cleanTo)) {
      const len = cleanFrom.length;
      const start = BigInt(cleanFrom);
      const end = BigInt(cleanTo);
      const result: string[] = [];

      for (let curr = start; curr <= end; curr++) {
        result.push(curr.toString().padStart(len, '0'));
      }
      return result;
    }

    const fromMatch = cleanFrom.match(/^([A-Za-z0-9_-]*?)(\d+)$/)!;
    const toMatch = cleanTo.match(/^([A-Za-z0-9_-]*?)(\d+)$/)!;
    const prefix = fromMatch[1];
    const digitsFrom = fromMatch[2];
    const digitsTo = toMatch[2];
    const len = digitsFrom.length;

    const start = BigInt(digitsFrom);
    const end = BigInt(digitsTo);
    const result: string[] = [];

    for (let curr = start; curr <= end; curr++) {
      result.push(prefix + curr.toString().padStart(len, '0'));
    }
    return result;
  }
}
